File: src/processing/circuit_breaker.py
"""
Circuit breaker pattern for preventing cascading failures.
"""

from enum import Enum
from datetime import datetime, timedelta
from typing import Callable, Any, Optional, Dict
import logging

logger = logging.getLogger(__name__)


class CircuitState(Enum):
    """Circuit breaker states"""
    CLOSED = "closed"      # Normal operation
    OPEN = "open"          # Failing, reject calls
    HALF_OPEN = "half_open"  # Testing recovery


class CircuitOpenError(Exception):
    """Raised when circuit is open and rejecting calls"""
    pass


class CircuitBreaker:
    """
    Prevents cascading failures by breaking the circuit after repeated failures.

    The circuit breaker has three states:
    - CLOSED: Normal operation, requests pass through
    - OPEN: Too many failures, requests are rejected immediately
    - HALF_OPEN: Testing if the system has recovered
    """

    def __init__(
        self,
        failure_threshold: int = 3,
        timeout: int = 60,
        success_threshold: int = 2,
        name: str = "circuit"
    ):
        """
        Initialize circuit breaker.

        Args:
            failure_threshold: Number of failures before opening circuit
            timeout: Seconds to wait before attempting recovery
            success_threshold: Successes needed in HALF_OPEN to close circuit
            name: Name for logging
        """
        self.failure_threshold = failure_threshold
        self.timeout = timeout  # Seconds before attempting recovery
        self.success_threshold = success_threshold
        self.name = name

        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time: Optional[datetime] = None
        self.last_attempt_time: Optional[datetime] = None
        self.last_error: Optional[Exception] = None

    def call(self, func: Callable, *args, **kwargs) -> Any:
        """
        Execute function with circuit breaker protection.

        Args:
            func: Function to execute
            *args: Positional arguments for func
            **kwargs: Keyword arguments for func

        Returns:
            Result from func

        Raises:
            CircuitOpenError: If circuit is open
            Exception: Any exception from func
        """
        if self.state == CircuitState.OPEN:
            if self._should_attempt_recovery():
                self.state = CircuitState.HALF_OPEN
                self.success_count = 0
                logger.info(f"Circuit breaker '{self.name}' entering HALF_OPEN state")
            else:
                time_remaining = self._time_until_recovery()
                raise CircuitOpenError(
                    f"Circuit '{self.name}' open due to {self.failure_count} failures. "
                    f"Retry in {time_remaining:.0f} seconds"
                )

        try:
            # Attempt the call
            self.last_attempt_time = datetime.now()
            result = func(*args, **kwargs)
            self._on_success()
            return result

        except Exception as e:
            self._on_failure(e)
            raise

    def _should_attempt_recovery(self) -> bool:
        """Check if enough time has passed to attempt recovery"""
        if self.last_failure_time is None:
            return True

        elapsed = datetime.now() - self.last_failure_time
        return elapsed.total_seconds() >= self.timeout

    def _time_until_recovery(self) -> float:
        """Calculate seconds until recovery attempt"""
        if self.last_failure_time is None:
            return 0

        elapsed = datetime.now() - self.last_failure_time
        remaining = self.timeout - elapsed.total_seconds()
        return max(0, remaining)

    def _on_success(self):
        """Handle successful call"""
        if self.state == CircuitState.HALF_OPEN:
            self.success_count += 1
            logger.debug(f"Circuit '{self.name}' success in HALF_OPEN: {self.success_count}/{self.success_threshold}")

            if self.success_count >= self.success_threshold:
                logger.info(f"Circuit breaker '{self.name}' recovered, entering CLOSED state")
                self.state = CircuitState.CLOSED
                self.failure_count = 0
                self.success_count = 0
                self.last_error = None
        else:
            # Reset failure count on success in CLOSED state
            if self.failure_count > 0:
                logger.debug(f"Circuit '{self.name}' success in CLOSED, resetting failure count")
            self.failure_count = 0

    def _on_failure(self, error: Exception):
        """Handle failed call"""
        self.failure_count += 1
        self.last_failure_time = datetime.now()
        self.last_error = error

        if self.state == CircuitState.HALF_OPEN:
            logger.warning(f"Circuit breaker '{self.name}' recovery failed, reopening circuit")
            self.state = CircuitState.OPEN
            self.success_count = 0

        elif self.failure_count >= self.failure_threshold:
            logger.error(f"Circuit breaker '{self.name}' opening after {self.failure_count} failures: {error}")
            self.state = CircuitState.OPEN

        else:
            logger.warning(f"Circuit '{self.name}' failure {self.failure_count}/{self.failure_threshold}: {error}")

    def reset(self):
        """Manually reset the circuit breaker to closed state"""
        logger.info(f"Circuit breaker '{self.name}' manually reset")
        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time = None
        self.last_error = None

    @property
    def is_open(self) -> bool:
        """Check if circuit is currently open"""
        return self.state == CircuitState.OPEN

    @property
    def is_closed(self) -> bool:
        """Check if circuit is currently closed"""
        return self.state == CircuitState.CLOSED

    def get_status(self) -> Dict[str, Any]:
        """
        Get current circuit breaker status.

        Returns:
            Dictionary with status information
        """
        status = {
            'name': self.name,
            'state': self.state.value,
            'failure_count': self.failure_count,
            'success_count': self.success_count,
        }

        if self.last_failure_time:
            status['last_failure'] = self.last_failure_time.isoformat()
            if self.state == CircuitState.OPEN:
                status['recovery_in'] = self._time_until_recovery()

        if self.last_error:
            status['last_error'] = str(self.last_error)

        return status


class MultiCircuitBreaker:
    """
    Manages multiple circuit breakers for different operations.
    """

    def __init__(self):
        """Initialize multi-circuit breaker manager"""
        self.breakers: Dict[str, CircuitBreaker] = {}

    def add_breaker(
        self,
        name: str,
        failure_threshold: int = 3,
        timeout: int = 60,
        success_threshold: int = 2
    ) -> CircuitBreaker:
        """
        Add a new circuit breaker.

        Args:
            name: Name of the breaker
            failure_threshold: Failures before opening
            timeout: Recovery timeout in seconds
            success_threshold: Successes needed to close

        Returns:
            The created circuit breaker
        """
        breaker = CircuitBreaker(
            failure_threshold=failure_threshold,
            timeout=timeout,
            success_threshold=success_threshold,
            name=name
        )
        self.breakers[name] = breaker
        return breaker

    def get_breaker(self, name: str) -> Optional[CircuitBreaker]:
        """Get a circuit breaker by name"""
        return self.breakers.get(name)

    def call(self, breaker_name: str, func: Callable, *args, **kwargs) -> Any:
        """
        Call function through named circuit breaker.

        Args:
            breaker_name: Name of breaker to use
            func: Function to call
            *args: Function arguments
            **kwargs: Function keyword arguments

        Returns:
            Function result

        Raises:
            ValueError: If breaker doesn't exist
            CircuitOpenError: If circuit is open
        """
        breaker = self.breakers.get(breaker_name)
        if not breaker:
            raise ValueError(f"No circuit breaker named '{breaker_name}'")

        return breaker.call(func, *args, **kwargs)

    def get_status(self) -> Dict[str, Dict[str, Any]]:
        """Get status of all circuit breakers"""
        return {
            name: breaker.get_status()
            for name, breaker in self.breakers.items()
        }

    def reset_all(self):
        """Reset all circuit breakers"""
        for breaker in self.breakers.values():
            breaker.reset()

File: typescript_lib/src/weight-processor-lib/core/processing/circuit_breaker.ts
/**
 * Circuit breaker pattern for preventing cascading failures.
 */

/**
 * Circuit breaker states
 */
export enum CircuitState {
  CLOSED = "closed", // Normal operation
  OPEN = "open", // Failing, reject calls
  HALF_OPEN = "half_open", // Testing recovery
}

/**
 * Raised when circuit is open and rejecting calls
 */
export class CircuitOpenError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CircuitOpenError";
    Object.setPrototypeOf(this, CircuitOpenError.prototype);
  }
}

/**
 * Circuit breaker status information
 */
export interface CircuitBreakerStatus {
  name: string;
  state: string;
  failure_count: number;
  success_count: number;
  last_failure?: string;
  recovery_in?: number;
  last_error?: string;
}

/**
 * Prevents cascading failures by breaking the circuit after repeated failures.
 *
 * The circuit breaker has three states:
 * - CLOSED: Normal operation, requests pass through
 * - OPEN: Too many failures, requests are rejected immediately
 * - HALF_OPEN: Testing if the system has recovered
 */
export class CircuitBreaker {
  private failure_threshold: number;
  private timeout: number; // Seconds before attempting recovery
  private success_threshold: number;
  private name: string;

  private state: CircuitState;
  private failure_count: number;
  private success_count: number;
  private last_failure_time: Date | null;
  private last_attempt_time: Date | null;
  private last_error: Error | null;

  /**
   * Initialize circuit breaker.
   *
   * @param failure_threshold - Number of failures before opening circuit
   * @param timeout - Seconds to wait before attempting recovery
   * @param success_threshold - Successes needed in HALF_OPEN to close circuit
   * @param name - Name for logging
   */
  constructor(
    failure_threshold: number = 3,
    timeout: number = 60,
    success_threshold: number = 2,
    name: string = "circuit"
  ) {
    this.failure_threshold = failure_threshold;
    this.timeout = timeout;
    this.success_threshold = success_threshold;
    this.name = name;

    this.state = CircuitState.CLOSED;
    this.failure_count = 0;
    this.success_count = 0;
    this.last_failure_time = null;
    this.last_attempt_time = null;
    this.last_error = null;
  }

  /**
   * Execute function with circuit breaker protection.
   *
   * @param func - Function to execute
   * @param args - Arguments for func
   * @returns Result from func
   * @throws {CircuitOpenError} If circuit is open
   * @throws {Error} Any exception from func
   */
  call<T>(func: (...args: any[]) => T, ...args: any[]): T {
    if (this.state === CircuitState.OPEN) {
      if (this._should_attempt_recovery()) {
        this.state = CircuitState.HALF_OPEN;
        this.success_count = 0;
        console.info(`Circuit breaker '${this.name}' entering HALF_OPEN state`);
      } else {
        const time_remaining = this._time_until_recovery();
        throw new CircuitOpenError(
          `Circuit '${this.name}' open due to ${this.failure_count} failures. ` +
            `Retry in ${Math.round(time_remaining)} seconds`
        );
      }
    }

    try {
      // Attempt the call
      this.last_attempt_time = new Date();
      const result = func(...args);
      this._on_success();
      return result;
    } catch (e) {
      this._on_failure(e instanceof Error ? e : new Error(String(e)));
      throw e;
    }
  }

  /**
   * Check if enough time has passed to attempt recovery
   */
  private _should_attempt_recovery(): boolean {
    if (this.last_failure_time === null) {
      return true;
    }

    const elapsed = Date.now() - this.last_failure_time.getTime();
    return elapsed / 1000 >= this.timeout;
  }

  /**
   * Calculate seconds until recovery attempt
   */
  private _time_until_recovery(): number {
    if (this.last_failure_time === null) {
      return 0;
    }

    const elapsed = (Date.now() - this.last_failure_time.getTime()) / 1000;
    const remaining = this.timeout - elapsed;
    return Math.max(0, remaining);
  }

  /**
   * Handle successful call
   */
  private _on_success(): void {
    if (this.state === CircuitState.HALF_OPEN) {
      this.success_count += 1;
      console.debug(
        `Circuit '${this.name}' success in HALF_OPEN: ${this.success_count}/${this.success_threshold}`
      );

      if (this.success_count >= this.success_threshold) {
        console.info(
          `Circuit breaker '${this.name}' recovered, entering CLOSED state`
        );
        this.state = CircuitState.CLOSED;
        this.failure_count = 0;
        this.success_count = 0;
        this.last_error = null;
      }
    } else {
      // Reset failure count on success in CLOSED state
      if (this.failure_count > 0) {
        console.debug(
          `Circuit '${this.name}' success in CLOSED, resetting failure count`
        );
      }
      this.failure_count = 0;
    }
  }

  /**
   * Handle failed call
   */
  private _on_failure(error: Error): void {
    this.failure_count += 1;
    this.last_failure_time = new Date();
    this.last_error = error;

    if (this.state === CircuitState.HALF_OPEN) {
      console.warn(
        `Circuit breaker '${this.name}' recovery failed, reopening circuit`
      );
      this.state = CircuitState.OPEN;
      this.success_count = 0;
    } else if (this.failure_count >= this.failure_threshold) {
      console.error(
        `Circuit breaker '${this.name}' opening after ${this.failure_count} failures: ${error}`
      );
      this.state = CircuitState.OPEN;
    } else {
      console.warn(
        `Circuit '${this.name}' failure ${this.failure_count}/${this.failure_threshold}: ${error}`
      );
    }
  }

  /**
   * Manually reset the circuit breaker to closed state
   */
  reset(): void {
    console.info(`Circuit breaker '${this.name}' manually reset`);
    this.state = CircuitState.CLOSED;
    this.failure_count = 0;
    this.success_count = 0;
    this.last_failure_time = null;
    this.last_error = null;
  }

  /**
   * Check if circuit is currently open
   */
  get is_open(): boolean {
    return this.state === CircuitState.OPEN;
  }

  /**
   * Check if circuit is currently closed
   */
  get is_closed(): boolean {
    return this.state === CircuitState.CLOSED;
  }

  /**
   * Get current circuit breaker status.
   *
   * @returns Dictionary with status information
   */
  get_status(): CircuitBreakerStatus {
    const status: CircuitBreakerStatus = {
      name: this.name,
      state: this.state,
      failure_count: this.failure_count,
      success_count: this.success_count,
    };

    if (this.last_failure_time) {
      status.last_failure = this.last_failure_time.toISOString();
      if (this.state === CircuitState.OPEN) {
        status.recovery_in = this._time_until_recovery();
      }
    }

    if (this.last_error) {
      status.last_error = String(this.last_error);
    }

    return status;
  }
}

/**
 * Manages multiple circuit breakers for different operations.
 */
export class MultiCircuitBreaker {
  private breakers: Map<string, CircuitBreaker>;

  /**
   * Initialize multi-circuit breaker manager
   */
  constructor() {
    this.breakers = new Map<string, CircuitBreaker>();
  }

  /**
   * Add a new circuit breaker.
   *
   * @param name - Name of the breaker
   * @param failure_threshold - Failures before opening
   * @param timeout - Recovery timeout in seconds
   * @param success_threshold - Successes needed to close
   * @returns The created circuit breaker
   */
  add_breaker(
    name: string,
    failure_threshold: number = 3,
    timeout: number = 60,
    success_threshold: number = 2
  ): CircuitBreaker {
    const breaker = new CircuitBreaker(
      failure_threshold,
      timeout,
      success_threshold,
      name
    );
    this.breakers.set(name, breaker);
    return breaker;
  }

  /**
   * Get a circuit breaker by name
   */
  get_breaker(name: string): CircuitBreaker | undefined {
    return this.breakers.get(name);
  }

  /**
   * Call function through named circuit breaker.
   *
   * @param breaker_name - Name of breaker to use
   * @param func - Function to call
   * @param args - Function arguments
   * @returns Function result
   * @throws {Error} If breaker doesn't exist
   * @throws {CircuitOpenError} If circuit is open
   */
  call<T>(breaker_name: string, func: (...args: any[]) => T, ...args: any[]): T {
    const breaker = this.breakers.get(breaker_name);
    if (!breaker) {
      throw new Error(`No circuit breaker named '${breaker_name}'`);
    }

    return breaker.call(func, ...args);
  }

  /**
   * Get status of all circuit breakers
   */
  get_status(): Record<string, CircuitBreakerStatus> {
    const status: Record<string, CircuitBreakerStatus> = {};
    for (const [name, breaker] of this.breakers.entries()) {
      status[name] = breaker.get_status();
    }
    return status;
  }

  /**
   * Reset all circuit breakers
   */
  reset_all(): void {
    for (const breaker of this.breakers.values()) {
      breaker.reset();
    }
  }
}
